pkg/controller/bundle/bundle_unpacker.go

```diff
@@ -0,0 +1,388 @@
+			cond.Status = corev1.ConditionTrue
+			cond.Reason = "JobIncomplete"
+			cond.Message = "unpack job not completed"
+			cond.LastTransitionTime = &now
+			result.SetCondition(cond)
+		}
+
+		return
+	}
+
+	result.bundle, err = c.loader.Load(cm)
+	if err != nil {
+		// TODO: This may never be used depending on how the returned error is interpreted; consider removing.
+		if cond.Status != corev1.ConditionTrue {
+			cond.Status = corev1.ConditionTrue
+			cond.Reason = "ConfigMapNotLoaded"
+			cond.Message = "unpack configmap was not loaded"
+			cond.LastTransitionTime = &now
+			result.SetCondition(cond)
+		}
+
+		return
+	}
+
+	// A successful load should remove the pending condition
+	result.RemoveCondition(operatorsv1alpha1.BundleLookupPending)
+
+	return
+}
+
+func (c *configmapUnpacker) ensureConfigmap(namespace, name string) (cm *corev1.ConfigMap, err error) {
+	fresh := &corev1.ConfigMap{}
+	fresh.SetNamespace(namespace)
+	fresh.SetName(name)
+
+	cm, err = c.cmLister.ConfigMaps(fresh.GetNamespace()).Get(fresh.GetName())
+	if apierrors.IsNotFound(err) {
+		cm, err = c.client.CoreV1().ConfigMaps(fresh.GetNamespace()).Create(fresh)
+	}
+
+	return
+}
+
+func (c *configmapUnpacker) ensureJob(configmapNamespace, configmapName, bundlePath string) (job *batchv1.Job, err error) {
+	fresh := c.job(configmapNamespace, configmapName, bundlePath)
+	job, err = c.jobLister.Jobs(fresh.GetNamespace()).Get(fresh.GetName())
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			job, err = c.client.BatchV1().Jobs(fresh.GetNamespace()).Create(fresh)
+		}
+
+		return
+	}
+
+	if equality.Semantic.DeepDerivative(fresh.Spec, job.Spec) {
+		return
+	}
+
+	// TODO: Decide when to fail-out instead of deleting the job
+	err = c.client.BatchV1().Jobs(job.GetNamespace()).Delete(job.GetName(), &metav1.DeleteOptions{})
+	job = nil
+	return
+}
+
+func (c *configmapUnpacker) ensureRole(configmapNamespace, configmapName string) (role *rbacv1.Role, err error) {
+	rule := rbacv1.PolicyRule{
+		APIGroups: []string{
+			"",
+		},
+		Verbs: []string{
+			"create", "get", "update",
+		},
+		Resources: []string{
+			"configmaps",
+		},
+		ResourceNames: []string{
+			configmapName,
+		},
+	}
+	fresh := &rbacv1.Role{
+		Rules: []rbacv1.PolicyRule{rule},
+	}
+	fresh.SetNamespace(configmapNamespace)
+	fresh.SetName(c.rbacName)
+
+	role, err = c.roleLister.Roles(fresh.GetNamespace()).Get(fresh.GetName())
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			role, err = c.client.RbacV1().Roles(fresh.GetNamespace()).Create(fresh)
+		}
+
+		return
+	}
+
+	// Add the policy rule if necessary
+	for _, existing := range role.Rules {
+		if equality.Semantic.DeepDerivative(rule, existing) {
+			return
+		}
+	}
+	role.Rules = append(role.Rules, rule)
+
+	role, err = c.client.RbacV1().Roles(role.GetNamespace()).Update(role)
+
+	return
+}
+
+func (c *configmapUnpacker) ensureRoleBinding(roleNamespace string) (roleBinding *rbacv1.RoleBinding, err error) {
+	fresh := &rbacv1.RoleBinding{
```

Comment:
shouldn't these ensured resources have some reasonable ownerrefs?

Reply:
Okay. I'll add ownerrefs in the direction of Role/RoleBindings/Jobs -> ConfigMap. 

The ConfigMap itself is a bit more tricky since there can exist InstallPlans that effectively reference it across several namespaces. The sensible thing to me is to add an ownerref in the direction of ConfigMap -> CatalogSource, and then propagate an error condition on the BundleLookup if it's not present when ensuring the ConfigMap.

Reply:
> there can exist InstallPlans that effectively reference it across several namespaces

Can you describe the case where this happens? Won't everything be scoped to one namespace?

Also - ownerrefs may not be needed at all if OLM will perform the cleanup (which might be nice for the rbac; it's not really needed after the job runs)